From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andreas Dinter <andreas.dinter@systec-electronic.com>
Date: Thu, 16 Jan 2025 16:50:42 +0100
Subject: [PATCH] board/ti/am62x/sysworxx.c: Detect RAM size at runtime

Removing spl_perform_fixups and adjusting the Makefile is needed to
avoid compiler and linker errors. Supported devices do not use ECC RAM
so we should be fine.

Signed-off-by: Andreas Dinter <andreas.dinter@systec-electronic.com>
---
 board/ti/am62x/sysworxx.c | 50 +++++++++++++++++++++++++++++++--------
 board/ti/common/Makefile  |  2 ++
 2 files changed, 42 insertions(+), 10 deletions(-)

diff --git a/board/ti/am62x/sysworxx.c b/board/ti/am62x/sysworxx.c
index dba5a1ece05..7698b0e1f1f 100644
--- a/board/ti/am62x/sysworxx.c
+++ b/board/ti/am62x/sysworxx.c
@@ -23,6 +23,7 @@
 #include <asm/arch/k3-ddr.h>
 #include <asm/gpio.h>
 #include <linux/delay.h>
+#include <display_options.h>
 
 #include "../common/fdt_ops.h"
 
@@ -84,6 +85,45 @@ void set_dfu_alt_info(char *interface, char *devstr)
 }
 #endif
 
+/*
+ * RAM size is detected at runtime. Maximum RAM size is 2G.
+ * (this replaces defaults in board/ti/common/k3-ddr.c which read this
+ * information from DT)
+ * To test this check output during boot and/or with the `bdinfo` command.
+ */
+
+#define CFG_SYS_SDRAM_SIZE_MAX	SZ_2G
+
+int dram_init(void)
+{
+	s32 ret = 0;
+
+	ret = fdtdec_setup_mem_size_base_lowest();
+	if (ret) {
+		printf("Error setting up mem size and base. %d\n", ret);
+		return ret;
+	}
+
+	gd->ram_size = get_ram_size((long *)gd->ram_base, CFG_SYS_SDRAM_SIZE_MAX);
+	printf("Detected RAM size: ");
+	print_size(gd->ram_size, "\n");
+
+	ret = k3_mem_map_init();
+	if (ret) {
+		printf("Error setting up MMU table. %d\n", ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+int dram_init_banksize(void)
+{
+    gd->bd->bi_dram[0].start = (phys_addr_t) gd->ram_base;
+    gd->bd->bi_dram[0].size = (phys_size_t) gd->ram_size;
+    return 0;
+}
+
 int board_init(void)
 {
 	return 0;
@@ -150,16 +190,6 @@ void spl_board_init(void)
 		splash_display();
 
 }
-
-void spl_perform_fixups(struct spl_image_info *spl_image)
-{
-	if (IS_ENABLED(CONFIG_K3_DDRSS)) {
-		if (IS_ENABLED(CONFIG_K3_INLINE_ECC))
-			fixup_ddr_driver_for_ecc(spl_image);
-	} else {
-		fixup_memory_node(spl_image);
-	}
-}
 #endif
 
 #if defined(CONFIG_OF_BOARD_SETUP)
diff --git a/board/ti/common/Makefile b/board/ti/common/Makefile
index caf6b9fa8c1..ae50ab7bca1 100644
--- a/board/ti/common/Makefile
+++ b/board/ti/common/Makefile
@@ -4,4 +4,6 @@
 obj-${CONFIG_TI_I2C_BOARD_DETECT} += board_detect.o
 obj-${CONFIG_CMD_EXTENSION} += cape_detect.o
 obj-${CONFIG_OF_LIBFDT} += fdt_ops.o
+ifndef CONFIG_TARGET_SYSWORXX_AM62X_ADOPT
 obj-${CONFIG_ARCH_K3} += k3-ddr.o
+endif
-- 
2.50.1

